### DISCUSSION TITLE
Allow customization of validation error

### DISCUSSION THREAD

**Discussion by johanfleury** [fastapi/fastapi#9061] [Questions]
### Is your feature request related to a problem

I’d like to globally customize validation errors by changing (overriding) the `HTTPValidationError`/`ValidationError` and the `422` status code.

### The solution you would like

Something great would be adding some parameters to the `FastAPI` class:

```python
from fastapi import FastAPI

from .models import ValidationError

app = FastAPI(
    validation_error_status_code=422,
    validation_error_model=ValidationError
)
```

### Describe alternatives you've considered

What’s currently possible (to my knowledge) is adding an item with status code `422`, `4XX` or `default` in the `responses` dict, but this as to be done manually for **every** route that will perform validation. This also prevent changing the status code to a specific value (you can either stick with 422, or have something vague like `default` or `4XX`).

**Comment by phy25** [comment]
Does this work for you? https://fastapi.tiangolo.com/tutorial/handling-errors/#override-request-validation-exceptions

**Comment by johanfleury** [comment]
Well it works regarding what’s produced by the API, but it doesn’t regarding documentation.

**Comment by aidan-** [comment]
I have hit a similar issue previously and solved it by modifying the OpenAPI spec 'manually' like it's been suggested in some [previous issues](https://github.com/tiangolo/fastapi/issues/1217#issuecomment-613087066). 

However, I think it would be useful if you could add 'additional status codes' and/or 'additional responses' to a `APIRouter()` instance, which would add them to all of the routes automatically.

**Comment by akrejczinger** [comment]
Extending OpenAPI manually seems a bit complicated to me.
HTTP 422 is hardcoded in openapi/utils.py. I solved my issue for now by modifying the FastAPI source code itself, but a proper way to do this automatically using the exception handlers would be preferable.

**Comment by rudaporto** [comment]
Hey @tiangolo,

I would like to work on this feature.
Do you think that it's a good idea?
Any tips or concerns?

**Comment by Kludex** [comment]
The solution that would require less work and have bigger chances on being approved (based only on what I've seen here) is creating a flag to not add the 422 response in the swagger documentation. 

The custom exception handler already handles the first problem described.

**Comment by rudaporto** [comment]
> The solution that would require less work and have bigger chances on being approved (based only on what I've seen here) is creating a flag to not add the 422 response in the swagger documentation.
> 
> The custom exception handler already handles the first problem described.

@Kludex thanks for the quick response.

It works for me and I think this is the only blocker for proper customizing the RequestValidationError into a custom response format.

**Comment by ushu** [comment]
If all you errors follow the same schema, and provided you implemented the custom handler described in https://fastapi.tiangolo.com/tutorial/handling-errors/#override-request-validation-exceptions, you can just override the default schema in [`openapi/utils`](https://github.com/tiangolo/fastapi/blob/master/fastapi/openapi/utils.py).

For example if I take the example server from the docs:

```python
from typing import Optional
from fastapi import FastAPI

app = FastAPI()

@app.get("/")
def read_root():
    return {"Hello": "World"}

@app.get("/items/{item_id}")
def read_item(item_id: int, q: Optional[str] = None):
    return {"item_id": item_id, "q": q}
```

it will present the default schema il http://localhost:8000/docs:

<img width="1211" alt="image" src="https://user-images.githubusercontent.com/1015378/103926254-f667e100-5118-11eb-887b-1636b59ac867.png">

But if I just **overwrite the `validation_error_response_definition`** in the [`fastapi.openapi.utils`](https://github.com/tiangolo/fastapi/blob/master/fastapi/openapi/utils.py#L35-L44) module it changes the default schema used by OpenAPI:

```python
from typing import Optional
from fastapi import FastAPI

# import the module we want to modify
import fastapi.openapi.utils as fu

app = FastAPI()

# and override the schema
fu.validation_error_response_definition = {
    "title": "HTTPValidationError",
    "type": "object",
    "properties": {
        "error": {"title": "Message", "type": "string"}, 
    },
}

@app.get("/")
def read_root():
    return {"Hello": "World"}

@app.get("/items/{item_id}")
def read_item(item_id: int, q: Optional[str] = None):
    return {"item_id": item_id, "q": q}
```

gives:

<img width="1211" alt="image" src="https://user-images.githubusercontent.com/1015378/103926751-a5a4b800-5119-11eb-9ae0-e79ab4d01b4c.png">

**Comment by kissgyorgy** [comment]
I would like this feature too, because it took more time to make it work than I wanted to 😞

Here is a less hacky way to do it.
You can override the 422 response type globally when instantiating a FastAPI application with the `responses` parameter:
```python
app = FastAPI(
    exception_handlers=(RequestValidationError, validation_exception_handler),
    responses={
        422: {"content": change_422_schema({})},
     },
)
```

First, you need to implement the `RequestValidationError` handler:
```python

class Error(BaseModel):
    type: str
    detail: str
    location: Optional[List[str]] = None

# We only need this class for FastApi in the responses parameter for path decorators
class ErrorResponse(BaseModel):
    errors: List[Error]


def validation_exception_handler(request: Request, exc: RequestValidationError):
    """Format the pydantic ValidationErrors in a more human-readable way."""
    error_res = ErrorResponse(
        errors=[
            Error(type=err["type"], detail=err["msg"], location=err["loc"])
            for err in exc.errors()
        ]
    )
    return _make_json_response(status.HTTP_422_UNPROCESSABLE_ENTITY, error_res)
```

Then, you need to modify the `responses` dict `content`, which will make it directly to the OpenAPI schema, by changing the `"application/json" / "schema" / "$ref"` value to your own Error type name:

```python
from fastapi.openapi.constants import REF_PREFIX

def change_422_schema(content: dict):
    """Change the ValidationError's schema to make it consistent with the rest of our API."""
    new_content = content.copy()
    content_type = new_content.setdefault("application/json", {})
    content_type["schema"] = {"$ref": REF_PREFIX + ErrorResponse.__name__}
    return new_content
```

So the 422 responses will be in the proper format and the generated OpenAPI schema will show the exact model and there will be no leftover models from FastAPI.

**Comment by tupui** [comment]
I am doing something simpler. Just removing any ref to the error 422 in the generated OpenAPI dict:

```python
from fastapi import FastAPI
from fastapi.openapi.utils import get_openapi
from fastapi.exceptions import RequestValidationError

app = FastAPI()


@app.get("/items/")
async def read_items():
    return [{"name": "Foo"}]


@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request, exc):
    """Your custom validation exception."""
    message = 
    return JSONResponse(
        status_code=400,
        content={"message": f"Validation error: {exc}"}
    )


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi()

    # look for the error 422 and removes it
    for method in openapi_schema["paths"]:
        try:
            del openapi_schema["paths"][method]["post"]["responses"]["422"]
        except KeyError:
            pass

    app.openapi_schema = openapi_schema
    return app.openapi_schema


app.openapi = custom_openapi

```

**Comment by Acerinth** [comment]
As this is still the topic... For all of you, who, like me, just want to **change the status code** to 400, but **keep the default response schema** AND still have it in the OpenAPI Docs, without the default 422... :D Here it goes...

Firstly, you describe your Pydantic models, like the original (used only for docs):
```python
from pydantic import BaseModel
from typing import Optional, List

class Error(BaseModel):

    loc: Optional[List[str]] = None
    msg: str
    type: str

class ErrorResponse(BaseModel):
    detail: List[Error]
```

Then you use Pydantic model `ErrorResponse` in responses when defining your app:
```python
app = FastAPI(title="My API", description="Some description here",
                     responses={status.HTTP_400_BAD_REQUEST: {"model": ErrorResponse}})
```

Then you override the default exception handler (like in the [tutorial](https://fastapi.tiangolo.com/tutorial/handling-errors/#override-request-validation-exceptions)):
```python
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse
from fastapi.encoders import jsonable_encoder

@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request, exc):
    return JSONResponse(content=jsonable_encoder({"detail": exc.errors()}),
                        status_code=status.HTTP_400_BAD_REQUEST)
```

With this, you will have 400 normally documented for all your routes. 

If you want to remove all 422 codes from the documentation, you need to exchange the default OpenAPI schema, similar to how @tupui  wrote above, or according to the tutorial [here](https://fastapi.tiangolo.com/advanced/extending-openapi/). Therefore, **after** defining all your path routes, add the following:
```python
def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title=app.title,
        version=app.version,
        routes=app.routes
    )

    http_methods = ["post", "get", "put", "delete"]
    # look for the error 422 and removes it
    for method in openapi_schema["paths"]:
        for m in http_methods:
            try:
                del openapi_schema["paths"][method][m]["responses"]["422"]
            except KeyError:
                pass

    app.openapi_schema = openapi_schema
    return app.openapi_schema


app.openapi = custom_openapi
```

**However**, this works for the default API with no additional root paths. In order to make it work for such scenario, or when you want this to work only on the **mounted API**, you have to add `servers` to the `openapi_schema` as well, to have the proper routing in Swagger/Redoc. For example:
```python
def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title=app.title,
        version=app.version,
        routes=app.routes,
        servers=[{
            "url": "/subapi"
        }]
    )
    # the rest as above
```

P.S. If you really wanna be persistent, and remove even the `ValidationError` and `HTTPValidationError` from schemas, you can add the following to your `custom_openapi()` method:
```python
    for schema in list(openapi_schema["components"]["schemas"]):
        if schema == "HTTPValidationError" or schema == "ValidationError":
            del openapi_schema["components"]["schemas"][schema]
```

Hope this helps! 😄 

Thanks again @tiangolo for all the awesome work on the framework! 🚀

**Comment by gustavoarmelin** [comment]
In my application ovverhide the 422 error for 400, and overhide the responses errors for all api for the example: `message: { field_error: message_error}`

    def get_message(root: str, fields: list, error_message: str, return_message=None):
        return_message = return_message or {}
    
        if root not in return_message:
            return_message[root] = {}
    
        for i, field in enumerate(fields):
            if len(fields) == 1:
                return_message[root].update({field: error_message})
            else:
                new_fields = fields.copy()
                del(new_fields[i])
                response = get_message(
                    root=field, fields=new_fields,
                    error_message=error_message, return_message=return_message[root]
                )
                return_message[root].update(response)
                break
    
        return return_message

And the method bellow is the rewrite `exception_handlers={RequestValidationError: validation_bad_request_exception_handler}`


    async def validation_bad_request_exception_handler(request, exc):
        message = {}
        for x in exc.errors():
            fields = [entry for entry in x['loc'] if entry != 'body']
    
            if len(fields) == 0:
                message.update({'error': 'invalid request'})
            elif len(fields) == 1:
                message.update({fields[0]: x['msg']})
            else:
                # for dict / list with errors inside it
                field_root = fields[0]
                del(fields[0])
    
                message = get_message(
                    root=field_root,
                    fields=fields,
                    error_message=x['msg'],
                    return_message=message
                )
    
        return JSONResponse(
            status_code=400,
            content={
                "message": message,
            },
        )

Its work fine for me =D
It is just missing to change the documentation

**Comment by i026e** [comment]
> I would like this feature too, because it took more time to make it work than I wanted to 
> 
> Here is a less hacky way to do it.
> You can override the 422 response type globally when instantiating a FastAPI application with the `responses` parameter:
> 
> ```python
> app = FastAPI(
>     exception_handlers=(RequestValidationError, validation_exception_handler),
>     responses={
>         422: {"content": change_422_schema({})},
>      },
> )

> Then, you need to modify the `responses` dict `content`, which will make it directly to the OpenAPI schema, by changing the `"application/json" / "schema" / "$ref"` value to your own Error type name:
> 
> ```python
> from fastapi.openapi.constants import REF_PREFIX
> 
> def change_422_schema(content: dict):
>     """Change the ValidationError's schema to make it consistent with the rest of our API."""
>     new_content = content.copy()
>     content_type = new_content.setdefault("application/json", {})
>     content_type["schema"] = {"$ref": REF_PREFIX + ErrorResponse.__name__}
>     return new_content
> ```
> 
> So the 422 responses will be in the proper format and the generated OpenAPI schema will show the exact model and there will be no leftover models from FastAPI.

I did similar for a single router but slightly simpler
```
app = FastAPI(...)
app.include_router(
        api_router,
        prefix=settings.API_V1_STR,
        responses={
            422: {
                "description": "Validation Error",
                "model": ErrorResponse
            }
        }
    )
```

**Comment by AlrasheedA** [comment]
Following up on @ushu excellent simple solution,
I would like to add that you could directly fetch the schema from your error model, ensuring consistent behavior between your response and its documentation. 

```python
from pydantic import BaseModel, Field
from fastapi import FastAPI


# define or import your error response model
class ErrorResponse(BaseModel):
    message: str = Field(..., example="message to the developer")
    error: str = Field(..., example="Validation Error")
    detail: dict = Field(..., example=[
        {'loc': 'quantity', 
         'msg': 'value is not a valid integer',
         'type':'type_error.integer'}])

# override fastapi 422 schema
import fastapi.openapi.utils as fu
fu.validation_error_response_definition = ErrorResponse.schema()

app = FastAPI()
```

Gives: 

<img width="711" alt="result" src="https://user-images.githubusercontent.com/33544979/152665891-87427606-5b7a-4433-842d-f2864bf23315.png">

**Comment by xshapira** [comment]
Even though following some great examples here, it took me time to make it work as I wanted. I attempted to make it as simple as possible for newcomers (like me) to grasp. 

Initially, we should implement the `RequestValidationError` handler, like so:
```python
from typing import List

from fastapi import FastAPI, Request, status
from fastapi.encoders import jsonable_encoder
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse
from pydantic import BaseModel


class Error(BaseModel):
    message: str
    type: str


class ErrorResponse(BaseModel):
    detail: List[Error]


def validation_exception_handler(request: Request, exc: RequestValidationError):
    """
    Format the pydantic ValidationErrors in a more human-readable way.
    """
    errors = {
        "detail": [
            {
                "message": err["msg"],
                "type": err["type"],
            }
        ]
        for err in exc.errors()
    }
    error_res = ErrorResponse(**errors)
    return JSONResponse(
        content=jsonable_encoder(error_res.detail[0]),
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
    )

```

Then we have to override the 422 response when instantiating the FastAPI application with the `responses` and `exception_handlers` parameters, like so:
```python
app = FastAPI(
    exception_handlers={RequestValidationError: validation_exception_handler},
    responses={
        422: {
            "description": "Validation Error",
            "model": ErrorResponse,
        },
    },
)
```

And now it will present our new schema:

![Screen Shot 2022-08-27 at 10 37 12](https://user-images.githubusercontent.com/48856190/187020335-65544c2b-6f79-464c-a609-0832c42554fa.JPG)

![Screen Shot 2022-08-27 at 10 37 45](https://user-images.githubusercontent.com/48856190/187020354-d8b84a75-f2f6-4388-a03a-4c625a28bc0c.JPG)

**Comment by fredrikaverpil** [comment]
I'm setting a custom openapi schema (as described [here](https://github.com/tiangolo/fastapi/issues/1376#issuecomment-828387176) and this works fine when I export the openapi schema as JSON and feed it into e.g. `redoc-cli` which builds the `redoc-static.html`.

But these customizations to the openapi schema are not reflected in FastAPI's served docs (when I access e.g. `http://localhost/redoc`). Is there any way I can have the cake and eat it (have the customizations show in both the exported openapi JSON and in the FastAPI-served docs)?
